README.md
# Golang RADIUS Protocol
Pure (as far as this code goes) implementation of RADIUS protocol in Go

Most of the RFCs related to RADIUS protocol are implemented with a few exceptions for a few data types, ie TLV, VSA

#### Go Version Support
![Go version](https://img.shields.io/badge/Go-1.20-brightgreen.svg)
![Go version](https://img.shields.io/badge/Go-1.19-brightgreen.svg)
![Go version](https://img.shields.io/badge/Go-1.18-brightgreen.svg)

Older versions may also be fine, however Action CI/CD only tests against versions above


## Installation
```
go get "github.com/MikhailMS/go-radius"

# In Go code
import "github.com/MikhailMS/go-radius"
```

## Tests
1. `go test ./...`


## Contributing
Would you love to contribute to this project? I'd really appreciate your input!

1. Raise an issue, if you believe some functionality is missing, something is broken and etc
2. Create a PR, if you already sorted out some issues yourself. **Please ensure** to submit PR to **development branch**


## Notes
1. This is re-write of [Rust implementation of RADIUS protocol](https://github.com/MikhailMS/rust-radius)


CHANGELOG.md
=============
# v0.2.0 (02 Jul 2023)

This is first complete release of `go-radius` package

## What's new
* `client` module:
    * Generic implementation of RadiusClient
* `protocol` module:
    * Implemented parse of dictionary files
    * Support for all core data types:
        * text
        * string
        * integer
        * integer64
        * time
        * ipaddr
        * ipv4addr
        * ipv4prefix
        * ipv6addr
        * ipv6prefix
        * ifid
    * Support for all Type Codes - as per `RFC 2865` & `RFC 3576`
    * All required functions to support work with Radius Packets
* `server` module:
    * Generic implementation of RadiusServer
* `tools` module:
    * Helpers to work with all core data types:
        * Integer to bytes and vice versa
        * Integer64 to bytes and vice versa
        * Time to bytes and vice versa
        * IPAddr, IPv4Addr & IPv4Prefix to bytes and vice versa
        * IPv6Addr & IPv6Prefix to bytes and vice versa
        * Ifid to bytes and vice versa
    * Helpers to encrypt/decrypt data (including salted)
* `examples` module:
    * Example for Radius Server
    * Example for Radius Client

## What's removed or deprecated

## What's changed


tools/tools_test.go
package tools

import (
  "testing"

  "github.com/stretchr/testify/assert"
)

func TestIPv6ToBytesWoSubnet(t *testing.T) {
  expectedBytes := []uint8{ 252, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 }
  
  ipv6Bytes, _ := IPv6StringToBytes("fc66::1")
  assert.Equal(t, expectedBytes, ipv6Bytes, "IPv6 bytes are not correct!")
}

func TestBytesToIPv6StringWoSubnet(t *testing.T) {
  expectedString := "fc66::1"

  ipv6Bytes := []uint8{ 252, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 }

  ipv6String, _ := BytesToIPv6String(ipv6Bytes)
  assert.Equal(t, expectedString, ipv6String, "IPv6 string is not correct!")
}

func TestIPv6ToBytesWSubnet(t *testing.T) {
  expectedBytes := []uint8{ 0, 64, 252, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 }

  ipv6Bytes, _ := IPv6StringToBytes("fc66::1/64")
  assert.Equal(t, expectedBytes, ipv6Bytes, "IPv6 bytes are not correct!")
}

func TestBytesToIPv6StringWSubnet(t *testing.T) {
  expectedString := "fc66::1/64"

  ipv6Bytes := []uint8{ 0, 64, 252, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 }

  ipv6String, _ := BytesToIPv6String(ipv6Bytes)
  assert.Equal(t, expectedString, ipv6String, "IPv6 string is not correct!")
}

func TestIPv4StringToBytes(t *testing.T) {
  expectedBytes := []uint8{ 192, 1, 10, 1 }

  ipv4Bytes, _ := IPv4StringToBytes("192.1.10.1")
  assert.Equal(t, expectedBytes, ipv4Bytes, "IPv4 bytes are not correct!")
}

func TestIPv4BytesToString(t *testing.T) {
  expectedString := "192.1.10.1"

  ipv4String, _ := BytesToIPv4String([]uint8{ 192, 1, 10, 1 })
  assert.Equal(t, expectedString, ipv4String, "IPv4 string is not correct!")
}

func TestIntegerToBytes(t *testing.T) {
  expectedBytes := []uint8{ 0, 0, 39, 16 }

  integer := uint32(10000)
  assert.Equal(t, expectedBytes, IntegerToBytes(integer), "Integer bytes is not correct!")
}

func TestBytesToInteger(t *testing.T) {
  expectedInteger := uint32(10000)

  integerBytes := []uint8{ 0, 0, 39, 16 }

  integer, _ := BytesToInteger(integerBytes)
  assert.Equal(t, expectedInteger, integer, "Integer is not correct!")
}

func TestInteger64ToBytes(t *testing.T) {
  expectedBytes := []uint8{ 0, 0, 0, 0, 0, 0, 39, 16 }

  integer := uint64(10000)
  assert.Equal(t, expectedBytes, Integer64ToBytes(integer), "Integer64 bytes is not correct!")
}

func TestBytesToInteger64(t *testing.T) {
  expectedInteger := uint64(10000)

  integerBytes := []uint8{ 0, 0, 0, 0, 0, 0, 39, 16 }

  integer, _ := BytesToInteger64(integerBytes)
  assert.Equal(t, expectedInteger, integer, "Integer64 is not correct!")
}

func TestTimestampToBytes(t *testing.T) {
  expectedBytes := []uint8 { 95, 71, 138, 29 }

  timestamp := int64(1598523933)

  timestampBytes, _ := TimestampToBytes(timestamp)
  assert.Equal(t, expectedBytes, timestampBytes, "Timestamp bytes is not correct!")
}

func TestBytesToTimestamp(t *testing.T) {
  expectedTimestamp := uint32(1598523933)

  timestampBytes := []uint8 { 95, 71, 138, 29 }

  timestamp, _ := BytesToTimestamp(timestampBytes)
  assert.Equal(t, expectedTimestamp, timestamp, "Timestamp integer is not correct!")
}


func TestEncryptData(t *testing.T) {
  expectedBytes := []uint8{ 135, 116, 155, 239, 226, 89, 90, 221, 62, 29, 218, 130, 102, 174, 191, 250 }

  secret        := []uint8("secret")
  data          := []uint8("password")
  authenticator := []uint8{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }

  encryptedBytes := EncryptData(&data, &authenticator, &secret)
  assert.Equal(t, expectedBytes, encryptedBytes, "Encrypted bytes are not correct!")
}

func TestEncryptDataLong(t *testing.T) {
  expectedBytes := []uint8{ 150, 53, 158, 249, 231, 79, 8, 213, 81, 115, 189, 162, 22, 207, 204, 137, 193,
                            149, 82, 147, 72, 149, 79, 48, 187, 199, 194, 200, 246, 6, 186, 182, 220, 19, 227, 32, 26, 20, 9, 152,
                            63, 40, 41, 91, 212, 22, 158, 54, 91, 247, 151, 67, 250,170, 105, 94, 20, 105, 120, 196, 237, 191, 99, 69 }

  secret        := []uint8("secret")
  data          := []uint8("a very long password, which will need multiple iterations")
  authenticator := []uint8{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }

  encryptedBytes := EncryptData(&data, &authenticator, &secret)
  assert.Equal(t, expectedBytes, encryptedBytes, "Encrypted bytes are not correct!")
}

func TestEncryptDataLimitLong(t *testing.T) {
  expectedBytes := []uint8{ 150, 53, 158, 249, 231, 79, 8, 213, 81, 115, 189, 162, 22, 207, 204, 137, 193, 149, 82, 147, 72, 149, 79, 48, 187, 199, 194, 200,
                            246, 6, 186, 182, 220, 19, 227, 32, 26, 20, 9, 152, 63, 40, 41, 91, 212, 22, 158, 54, 91, 247, 151, 67, 250, 170, 105, 94, 20, 71,
                            88, 165, 205, 201, 6, 55, 222, 205, 192, 227, 172, 93, 166, 15, 33, 86, 56, 181, 52, 4, 49, 190, 186, 17, 125, 50, 140, 52, 130, 194,
                            125, 93, 177, 65, 217, 195, 23, 75, 175, 219, 244, 156, 133, 145, 20, 176, 36, 90, 16, 77, 148, 221, 251, 155, 9, 107, 213, 140, 107,
                            112, 161, 99, 6, 108, 106, 33, 69, 192, 191, 98, 30, 147, 197, 72, 160, 234, 50, 243, 195, 62, 72, 225, 19, 63, 28, 221, 164, 43, 67,
                            63, 206, 208, 124, 254, 202, 118, 229, 58, 180, 210, 100, 149, 120, 97, 23, 203, 197, 139, 244, 241, 175, 232, 149, 77, 43, 231, 27, 56,
                            250, 58, 251, 6, 203, 197, 190, 78, 83, 127, 164, 31, 211, 52, 74, 92, 36, 250, 236, 210, 72, 52, 55, 248, 161, 160, 95, 102, 63, 190, 43,
                            253, 224, 114, 62, 23, 11, 242, 186, 91, 132, 14, 76, 171, 26, 1, 51, 78, 144, 50, 228, 212, 47, 104, 98, 60, 245, 1, 103, 217, 49, 105,
                            38, 108, 93, 85, 224, 227, 33, 50, 144, 0, 233, 54, 174, 67, 174, 101, 189, 41 }

  secret        := []uint8("secret")
  data          := []uint8("a very long password, which will need multiple iterations. a very long password, which will need multiple iterations. a very long password, which will need multiple iterations. a very long password, which will need multiple iterations. a very long passw")
  authenticator := []uint8{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }

  encryptedBytes := EncryptData(&data, &authenticator, &secret)
  assert.Equal(t, expectedBytes, encryptedBytes, "Encrypted bytes are not correct!")
}


func TestDecryptData(t *testing.T) {
  expectedData  := []uint8("password")
  
  secret        := []uint8("secret")
  authenticator := []uint8{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }
  encryptedData := []uint8{ 135, 116, 155, 239, 226, 89, 90, 221, 62, 29, 218, 130, 102, 174, 191, 250 }

  decryptedData := DecryptData(&encryptedData, &authenticator, &secret)
  assert.Equal(t, expectedData, decryptedData, "Decrypted data is not correct!")
}

func TestDecryptDataLong(t *testing.T) {
  expectedData  := []uint8("a very long password, which will need multiple iterations")
  
  secret        := []uint8("secret")
  authenticator := []uint8{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }

  encryptedData := []uint8{ 150, 53, 158, 249, 231, 79, 8, 213, 81, 115, 189, 162, 22, 207, 204, 137, 193,
                            149, 82, 147, 72, 149, 79, 48, 187, 199, 194, 200, 246, 6, 186, 182, 220, 19, 227, 32, 26, 20, 9, 152, 63,
                            40, 41, 91, 212, 22, 158, 54, 91, 247, 151, 67, 250,170, 105, 94, 20, 105, 120, 196, 237, 191, 99, 69 }

  decryptedData := DecryptData(&encryptedData, &authenticator, &secret)
  assert.Equal(t, expectedData, decryptedData, "Decrypted data is not correct!")
}

func TestDecryptDataLimitLong(t *testing.T) {
  expectedData  := []uint8("a very long password, which will need multiple iterations. a very long password, which will need multiple iterations. a very long password, which will need multiple iterations. a very long password, which will need multiple iterations. a very long passw")
  
  secret        := []uint8("secret")
  authenticator := []uint8{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }

  encryptedData := []uint8{ 150, 53, 158, 249, 231, 79, 8, 213, 81, 115, 189, 162, 22, 207, 204, 137, 193, 149, 82, 147, 72, 149, 79, 48, 187, 199, 194, 200,
                            246, 6, 186, 182, 220, 19, 227, 32, 26, 20, 9, 152, 63, 40, 41, 91, 212, 22, 158, 54, 91, 247, 151, 67, 250, 170, 105, 94, 20, 71,
                            88, 165, 205, 201, 6, 55, 222, 205, 192, 227, 172, 93, 166, 15, 33, 86, 56, 181, 52, 4, 49, 190, 186, 17, 125, 50, 140, 52, 130, 194,
                            125, 93, 177, 65, 217, 195, 23, 75, 175, 219, 244, 156, 133, 145, 20, 176, 36, 90, 16, 77, 148, 221, 251, 155, 9, 107, 213, 140, 107,
                            112, 161, 99, 6, 108, 106, 33, 69, 192, 191, 98, 30, 147, 197, 72, 160, 234, 50, 243, 195, 62, 72, 225, 19, 63, 28, 221, 164, 43, 67,
                            63, 206, 208, 124, 254, 202, 118, 229, 58, 180, 210, 100, 149, 120, 97, 23, 203, 197, 139, 244, 241, 175, 232, 149, 77, 43, 231, 27, 56,
                            250, 58, 251, 6, 203, 197, 190, 78, 83, 127, 164, 31, 211, 52, 74, 92, 36, 250, 236, 210, 72, 52, 55, 248, 161, 160, 95, 102, 63, 190, 43,
                            253, 224, 114, 62, 23, 11, 242, 186, 91, 132, 14, 76, 171, 26, 1, 51, 78, 144, 50, 228, 212, 47, 104, 98, 60, 245, 1, 103, 217, 49, 105,
                            38, 108, 93, 85, 224, 227, 33, 50, 144, 0, 233, 54, 174, 67, 174, 101, 189, 41 }

  decryptedData := DecryptData(&encryptedData, &authenticator, &secret)
  assert.Equal(t, expectedData, decryptedData, "Decrypted data is not correct!")
}

func TestSaltEncryptData(t *testing.T) {
  encryptedData := []uint8{ 0x85, 0x9a, 0xe3, 0x88, 0x34, 0x49, 0xf2, 0x1e, 0x14, 0x4c, 0x76, 0xc8, 0xb2, 0x1a, 0x1d, 0x4f, 0x0c, 0xdc }

  secret        := []uint8("secret")
  plaintext     := []uint8("password")
  salt          := encryptedData[:2]
  authenticator := make([]uint8, 16)

  assert.Equal(t, encryptedData, SaltEncryptData(&plaintext, &authenticator, &salt, &secret), "SaltEncryptData data is not correct!")
}

func TestSaltEncryptDataLongData(t *testing.T) {
  encryptedData := []uint8{ 0x85, 0xd9, 0x61, 0x72, 0x75, 0x37, 0xcf, 0x15, 0x20,
    0x19, 0x3b, 0x38, 0x39, 0x0e, 0x42, 0x21, 0x9b, 0x5e, 0xcb, 0x93, 0x25, 0x7d, 0xb4, 0x07,
    0x0c, 0xc1, 0x52, 0xcf, 0x38, 0x76, 0x29, 0x02, 0xc7, 0xb1, 0x29, 0xdf, 0x63, 0x96, 0x26,
    0x1a, 0x27, 0xe5, 0xc3, 0x13, 0x78, 0xa7, 0x97, 0xd8, 0x97, 0x9a, 0x45, 0xc3, 0x70, 0xd3,
    0xe4, 0xe2, 0xae, 0xd0, 0x55, 0x77, 0x19, 0xa5, 0xb6, 0x44, 0xe6, 0x8a }
  
  secret        := []uint8("secret")
  authenticator := make([]uint8, 16)
  plaintext     := []uint8("a very long password, which will need multiple iterations")
  salt          := encryptedData[:2]

  assert.Equal(t, encryptedData, SaltEncryptData(&plaintext, &authenticator, &salt, &secret), "SaltEncryptData data is not correct!")
}

func TestSaltDecryptData(t *testing.T) {
  encryptedData := []uint8{ 0x85, 0x9a, 0xe3, 0x88, 0x34, 0x49, 0xf2, 0x1e, 0x14, 0x4c, 0x76, 0xc8, 0xb2, 0x1a, 0x1d, 0x4f, 0x0c, 0xdc }
  
  secret        := []uint8("secret")
  authenticator := make([]uint8, 16)
  plaintext     := []uint8("password")

  decryptedData, _ := SaltDecryptData(&encryptedData, &authenticator, &secret)
  assert.Equal(t, plaintext, decryptedData, "SaltDecryptData data is not correct!")
}

func TestSaltDecryptDataLongData(t *testing.T) {
  encryptedData := []uint8{ 0x85, 0xd9, 0x61, 0x72, 0x75, 0x37, 0xcf, 0x15, 0x20,
    0x19, 0x3b, 0x38, 0x39, 0x0e, 0x42, 0x21, 0x9b, 0x5e, 0xcb, 0x93, 0x25, 0x7d, 0xb4, 0x07,
    0x0c, 0xc1, 0x52, 0xcf, 0x38, 0x76, 0x29, 0x02, 0xc7, 0xb1, 0x29, 0xdf, 0x63, 0x96, 0x26,
    0x1a, 0x27, 0xe5, 0xc3, 0x13, 0x78, 0xa7, 0x97, 0xd8, 0x97, 0x9a, 0x45, 0xc3, 0x70, 0xd3,
    0xe4, 0xe2, 0xae, 0xd0, 0x55, 0x77, 0x19, 0xa5, 0xb6, 0x44, 0xe6, 0x8a }
  
  secret        := []uint8("secret")
  authenticator := make([]uint8, 16)
  plaintext     := []uint8("a very long password, which will need multiple iterations")

  decryptedData, _ := SaltDecryptData(&encryptedData, &authenticator, &secret)
  assert.Equal(t, plaintext, decryptedData, "SaltDecryptData data is not correct!")
}


protocol/radius_packet.go
// RADIUS Packet implementation
package protocol

import (
  "errors"
  "fmt"
  "log"

  "crypto/hmac"
  "crypto/md5"
  "encoding/binary"
  "math/rand"
  "unicode/utf8"

  "github.com/MikhailMS/go-radius/tools"
)

// RadiusMsgType represents allowed types of RADIUS messages/packets
//
// Mainly used in RADIUS Server implementation to distinguish between sockets and functions, that should
// process RADIUS packets
type RadiusMsgType int

const (
  // Authentication packet
  AUTH RadiusMsgType = iota
  // Accounting packet
  ACCT
  // Change of Authorisation packet
  COA
)

// TypeCode represents all supported Codes of RADIUS message/packet
// as defined in RFC 2865 & RFC 3576
type TypeCode int

const (
  // AccessRequest      = 1
  AccessRequest TypeCode = iota
  // AccessAccept       = 2
  AccessAccept
  // AccessReject       = 3
  AccessReject
  // AccountingRequest  = 4
  AccountingRequest
  // AccountingResponse = 5
  AccountingResponse
  // AccessChallenge    = 11
  AccessChallenge
  // StatusServer       = 12
  StatusServer
  // StatusClient       = 13
  StatusClient
  // DisconnectRequest  = 40
  DisconnectRequest
  // DisconnectACK      = 41
  DisconnectACK
  // DisconnectNAK      = 42
  DisconnectNAK
  // CoARequest         = 43
  CoARequest
  // CoAACK             = 44
  CoAACK
  // CoANAK             = 45
  CoANAK
)

func typeCodeFromUint8(code uint8) (TypeCode, bool) {
  switch code {
    case 1:
      return AccessRequest, true
    case 2:
      return AccessAccept, true
    case 3:
      return AccessReject, true
    case 4:
      return AccountingRequest, true
    case 5:
      return AccountingResponse, true
    case 11:
      return AccessChallenge, true
    case 12:
      return StatusServer, true
    case 13:
      return StatusClient, true
    case 40:
      return DisconnectRequest, true
    case 41:
      return DisconnectACK, true
    case 42:
      return DisconnectNAK, true
    case 43:
      return CoARequest, true
    case 44:
      return CoAACK, true
    case 45:
      return CoANAK, true
    default:
      return 0, false
  }
}

func typeCodeToUint8(code TypeCode) (uint8, bool) {
  switch code {
    case AccessRequest:
      return 1, true
    case AccessAccept:
      return 2, true
    case AccessReject:
      return  3, true
    case AccountingRequest:
      return  4, true
    case AccountingResponse:
      return 5, true
    case AccessChallenge:
      return 11, true
    case StatusServer:
      return 12, true
    case StatusClient:
      return 13, true
    case DisconnectRequest:
      return 40, true
    case DisconnectACK:
      return 41, true
    case DisconnectNAK:
      return 42, true
    case CoARequest:
      return 43, true
    case CoAACK:
      return 44, true
    case CoANAK:
      return 45, true
    default:
      return 0, false
  }
}

// RadiusAttribute represents an attribute, which would be sent to RADIUS Server/client as a part of RadiusPacket
type RadiusAttribute struct {
  id    uint8
  name  string
  value []uint8
}

// CreateRadAttributeByName creates RadiusAttribute with given name
//
// Returns nil if ATTRIBUTE with such name is not found in Dictionary
func CreateRadAttributeByName(dictionary *Dictionary, attributeName string, value *[]uint8) (RadiusAttribute, bool) {
  for _, attr := range dictionary.Attributes() {
    if attr.Name() == attributeName {
      return RadiusAttribute {attr.Code(), attributeName, *value}, true
    }
  }

  return RadiusAttribute{}, false
}

// CreateRadAttributeByID creates RadiusAttribute with given id
//
// Returns nil if ATTRIBUTE with such id is not found in Dictionary
func CreateRadAttributeByID(dictionary *Dictionary, attributeID uint8, value *[]uint8) (RadiusAttribute, bool) {
  for _, attr := range dictionary.Attributes() {
    if attr.Code() == attributeID {
      return RadiusAttribute {attributeID, attr.Name(), *value}, true
    }
  }

  return RadiusAttribute{}, false
}

// OverrideValue overriddes RadiusAttribute value
//
// Mainly used when building Message-Authenticator
func (radAttr *RadiusAttribute) OverrideValue(newValue []uint8) {
  radAttr.value = newValue
}

// ID returns RadiusAttribute id
func (radAttr *RadiusAttribute) ID() uint8 {
  return radAttr.id
}

// Value returns RadiusAttribute value
func (radAttr *RadiusAttribute) Value() []uint8 {
  return radAttr.value
}

// Name returns RadiusAttribute name
func (radAttr *RadiusAttribute) Name() string {
  return radAttr.name
}

// VerifyOriginalValue verifies RadiusAttribute value, based on the ATTRIBUTE code type
func (radAttr *RadiusAttribute) VerifyOriginalValue(allowedType SupportedAttributeTypes) bool {
  switch allowedType {
    case AsciiString:
      return utf8.Valid(radAttr.value)
    case ByteString:
      if string(radAttr.value) != "" {
        return true
      }
      return false
    case Integer:
      _, ok := tools.BytesToInteger(radAttr.value)
      if ok {
        return true
      }
      return false
    case Date:
      _, ok := tools.BytesToTimestamp(radAttr.value)
      if ok {
        return true
      }
      return false
    case IPv4Addr:
      value, _ := tools.BytesToIPv4String(radAttr.value)
      if value  != "" {
        return true
      }
      return false
    case IPv6Addr:
      _, ok := tools.BytesToIPv6String(radAttr.value)
      if ok {
        return true
      }
      return false
    case IPv6Prefix:
      _, ok := tools.BytesToIPv6String(radAttr.value)
      if ok {
        return true
      }
      return false
    default:
      return false
  }
}

// OriginalStringValue returns RadiusAttribute value, if the attribute is dictionary's ATTRIBUTE with code type string, ipaddr,
// ipv6addr or ipv6prefix
func (radAttr *RadiusAttribute) OriginalStringValue(allowedType SupportedAttributeTypes) (string, bool) {
  switch allowedType {
    case AsciiString:
      ok := utf8.Valid(radAttr.value)
      if !ok {
        return "", false
      }
      return string(radAttr.value), true
    case IPv4Addr:
      value, err := tools.BytesToIPv4String(radAttr.value)
      if err != nil {
        return "", false
      }
      return value, true
    case IPv6Addr:
      return tools.BytesToIPv6String(radAttr.value)
    case IPv6Prefix:
      return tools.BytesToIPv6String(radAttr.value)
    default:
      return "", false
  }
}

// OriginalIntegerValue returns RadiusAttribute value, if the attribute is dictionary's ATTRIBUTE with code type
// integer or date
func (radAttr *RadiusAttribute) OriginalIntegerValue(allowedType SupportedAttributeTypes) (uint32, bool) {
  switch allowedType {
    case Integer:
      return tools.BytesToInteger(radAttr.value)
    case Date:
      return tools.BytesToTimestamp(radAttr.value)
    default:
      return 0, false
  }
}

// toBytes converts RadiusAttribute into ready-to-be-sent bytes slice
func (radAttr *RadiusAttribute) toBytes() []uint8 {
  /*
   *
   *         0               1              2
      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
     |     Type      |    Length     |  Value ...
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
  *  Taken from https://tools.ietf.org/html/rfc2865#page-23
  */
  var output []uint8

  output = append(output, radAttr.id)
  output = append(output, uint8(2 + len(radAttr.value)))
  output = append(output, radAttr.value...)

  return output
}


// RadiusPacket represents RADIUS packet
type RadiusPacket struct {
  id            uint8
  code          TypeCode
  authenticator []uint8
  attributes    []RadiusAttribute
}

// InitialisePacket initialises RADIUS packet with random ID and authenticator
func InitialiseRadiusPacket(code TypeCode) RadiusPacket {
  return RadiusPacket {createPacketId(), code, createPacketAuthenticator(), []RadiusAttribute{}}
}

// InitialisePacketFromBytes initialises RADIUS packet from raw bytes
func InitialiseRadiusPacketFromBytes(dictionary *Dictionary, bytes *[]uint8) (RadiusPacket, error) {
  var attributes []RadiusAttribute
 
  code, ok := typeCodeFromUint8((*bytes)[0])
  if !ok {
    return RadiusPacket{}, errors.New("Invalid TypeCode")
  }
  id   := (*bytes)[1]
  authenticator := (*bytes)[4:20]

  lastIndex := 20

  for {
    if lastIndex == len(*bytes) { break }
    
    attrID     := (*bytes)[lastIndex]
    attrLength := int((*bytes)[lastIndex + 1])
    attrValue  := (*bytes)[(lastIndex + 2):(lastIndex + attrLength)]

    _tmpAttr, ok := CreateRadAttributeByID(dictionary, attrID, &attrValue)
    if !ok {
      return RadiusPacket{}, errors.New(fmt.Sprintf("attribute with ID: %d is not found in dictionary", attrID))
    }
    attributes = append(attributes, _tmpAttr)
    lastIndex += attrLength
  }

  return RadiusPacket {id, code, authenticator, attributes}, nil
}

// SetAttributes sets attrbiutes for RadiusPacket
func (radPacket *RadiusPacket) SetAttributes(attr []RadiusAttribute) {
  radPacket.attributes = attr
}

// Overrides RadiusPacket id
func (radPacket *RadiusPacket) OverrideID(id uint8) {
  radPacket.id = id
}

// Overrides RadiusPacket authenticator
func (radPacket *RadiusPacket) OverrideAuthenticator(authenticator []uint8) {
  radPacket.authenticator = authenticator
}

// Overrides RadiusPacket Message-Authenticator
//
// Note: would fail if RadiusPacket has no Message-Authenticator attribute defined
func (radPacket *RadiusPacket) OverrideMessageAuthenticator(newMessageAuth []uint8) error {
  for idx := range radPacket.attributes {
    attr := &radPacket.attributes[idx]
    if attr.Name() == "Message-Authenticator" {
      attr.OverrideValue(newMessageAuth)
      return nil
    }
  }

  return errors.New("Message-Authenticator attribute not found in packet")
}

// Generates HMAC-MD5 hash for Message-Authenticator attribute
//
// Note 1: this function assumes that RadiusAttribute Message-Authenticator already exists in RadiusPacket
// Note 2: Message-Authenticator in RadiusPacket would be overwritten when this function is called
func (radPacket *RadiusPacket) GenerateMessageAuthenticator(secret string) error {
  // Step 1. Set Message-Authenticator to an array of 16 zeros in the RadiusPacket
  zeroedAuthenticator := make([]uint8, 16)

  err := radPacket.OverrideMessageAuthenticator(zeroedAuthenticator)
  if err != nil {
    return err
  }

  // Step 2. Calculate HMAC-MD5 for the entire RadiusPacket
  packetBytes, ok := radPacket.ToBytes()
  if !ok {
    return errors.New("failed to convert RadiusPacket to bytes")
  }

  hash := hmac.New(md5.New, []uint8(secret))
  hash.Write(packetBytes)

  // Step 3. Set Message-Authenticator to the result of Step 2
  err = radPacket.OverrideMessageAuthenticator(hash.Sum(nil))
  if err != nil {
    return err
  }

  return nil
}

// MessageAuthenticator returns Message-Authenticator value, if exists in RadiusPacket
// otherwise returns an error
func (radPacket *RadiusPacket) MessageAuthenticator() ([]uint8, error) {
  for _, attr := range radPacket.attributes {
    if attr.Name() == "Message-Authenticator" {
      return attr.value, nil
    }
  }

  return nil, errors.New("Message-Authenticator attribute not found in packet")
}

// ID returns RadiusPacket id
func (radPacket *RadiusPacket) ID() uint8 {
  return radPacket.id
}

// Authenticator returns RadiusPacket authenticator
func (radPacket *RadiusPacket) Authenticator() []uint8 {
  return radPacket.authenticator
}

// Code returns RadiusPacket code
func (radPacket *RadiusPacket) Code() TypeCode {
  return radPacket.code
}

// Attributes returns RadiusPacket attributes
func (radPacket *RadiusPacket) Attributes() []RadiusAttribute {
  return radPacket.attributes
}

// AttributeByName returns RadiusAttribute with given name
func (radPacket *RadiusPacket) AttributeByName(attrName string) RadiusAttribute {
  for _, attr := range radPacket.attributes {
    if attr.Name() == attrName {
      return attr
    }
  }

  return RadiusAttribute{}
}

// AttributeByID returns RadiusAttribute with given id
func (radPacket *RadiusPacket) AttributeByID(attrID uint8) RadiusAttribute {
  for _, attr := range radPacket.attributes {
    if attr.ID() == attrID {
      return attr
    }
  }

  return RadiusAttribute{}
}

// ToBytes converts RadiusPacket into ready-to-be-sent bytes slice
func (radPacket *RadiusPacket) ToBytes() ([]uint8, bool) {
  /* Prepare packet for a transmission to server/client
   *
   *          0               1               2         3
      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     |     Code      |  Identifier   |            Length             |
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     |                                                               |
     |                         Authenticator                         |
     |                                                               |
     |                                                               |
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     |  Attributes ...
     +-+-+-+-+-+-+-+-+-+-+-+-+-
   * Taken from https://tools.ietf.org/html/rfc2865#page-14
   *
   */
  var packetBytes []uint8
  var packetAttr  []uint8

  if len(radPacket.authenticator) == 0 {
    radPacket.authenticator = createPacketAuthenticator()
  }

  for _, attr := range radPacket.attributes {
    packetAttr = append(packetAttr, attr.toBytes()...)
  }

  code, ok := typeCodeToUint8(radPacket.code)
  if !ok {
    log.Println("WARNING: encountered invalid TypeCode when converting RadiusPacket to bytes")
    return []uint8{}, false
  }
  packetBytes = append(packetBytes, code)
  packetBytes = append(packetBytes, radPacket.id)
  packetBytes = append(packetBytes, packetLengthToBytes(uint16(20 + len(packetAttr)))...)
  packetBytes = append(packetBytes, radPacket.authenticator...)
  packetBytes = append(packetBytes, packetAttr...)

  return packetBytes, true
}


// createPacketId creates random uint8 ID for RadiusPacket
func createPacketId() uint8 {
  return uint8(rand.Intn(256))
}

// createPacketAuthenticator creates an uint8 slice of length 16
// filled with random numbers
func createPacketAuthenticator() []uint8 {
  var authenticator []uint8

  for i := 0; i < 16; i++ {
    authenticator = append(authenticator, uint8(rand.Intn(256)))
  }

  return authenticator
}

// packetLengthToBytes converts uint16 into []uint8 (of length 2)
func packetLengthToBytes(length uint16) []uint8 {
  bytes := make([]byte, 2)

  binary.BigEndian.PutUint16(bytes, length)
  return bytes
}
